integrated_channels/canvas/client.py

```diff
@@ -117,8 +117,70 @@ def delete_content_metadata(self, serialized_data):
 
     def create_assessment_reporting(self, user_id, payload):
         """
-        Not implemented yet
+        Send assessment level learner data, retrieved by the integrated channels exporter, to Canvas in the form of
+        an assignment and submission.
         """
+        learner_data = json.loads(payload)
+        self._create_session()
+
+        # Retrieve the Canvas user ID from the user's edx email (it is assumed that the learner's Edx
+        # and Canvas emails will match).
+        canvas_user_id = self._search_for_canvas_user_by_email(user_id)
+
+        # With the Canvas user ID, retrieve all courses for the user.
+        user_courses = self._get_canvas_user_courses_by_id(canvas_user_id)
+
+        # Find the course who's integration ID matches the learner data course ID
+        course_id = None
+        for course in user_courses:
+            integration_id = course['integration_id']
+            if integration_id == learner_data['courseID']:
+                course_id = course['id']
+                break
+
+        if not course_id:
+            raise ClientError(
```

Comment:
Is the learner_data passed into this method limited to the catalog(s) the customer is sending to their integration?

Ie, could we trip this error way too much due to us passing a bunch of courses in that are not even in their LMS because it's in a different catalog?

Reply:
> limited to the catalog(s) the customer is sending to their integration

The ultimate truth/source of `assessment level course learner data` being passed into this method comes from an enterprise enrollment giving us the course run ID as well as user ID. We use that information to query the gradebook view in platform which returns all the necessary learner data related to each subsection within the class. We then export and pass each subsection worth of data to the client which posts it to Canvas, one at a time.

So, to your question of is the learner data limited to the catalogs that the customer has designated? ...I would say no? I'm hesitant because while the data being passed in does not consider customer catalogs when retrieving it, (only the gradebook view which retrieves the user's grade data from our system), I don't think an enterprise learner would be enrolled in courses not contained within the customer's catalog... This obviously differs from how we export and post `course metadata` as that comes from the catalogs. Currently I am looking into and am slightly uneasy about our logic on posting both courses and course runs as it affects the way we match a learner data `course ID` to the Canvas' integration ID (as the learner data course ID will almost certainly always be a course run ID, where as I'm not entirely sure if the Canvas integration ID's will be course run ID's, course ID's or a mix of both). 

I can see the following cases (with varying degrees of possibilities of them actually happening IRL) causing course_id to not be found here:
1) Canvas changes their API response such that their `integration id` field no longer matches our expected set ID that we created when posting the course or there is a bug introduced into our code such that integration ID's don't match the gradebook view's `course ID` 
2) The user is enrolled in a Canvas course that has an edX `course ID`, rather than an edX `course run ID` as the integration ID, and our learner data `course ID` is the edX `course run ID`.
3) Something is horribly wrong and there just isn't a Canvas course that correlates to the edX course linked to the exported learner data and the enterprise learner has still somehow enrolled, started and completed a subsection of the edX course.

I don't think any of these are likely.. except for the `course ID` vs `course run ID` situation, but that is also dependent on how we/the Canvas admins are handling courses on their end. (Do they want one Canvas course to represent all edX courses, or do they one a Canvas course per course run that they're enrolling students in). This is a talk that I am trying to schedule with George and Ken to get better context on.

- [where we break down the **learner data export** by enrollment](https://github.com/edx/edx-enterprise/blob/master/integrated_channels/integrated_channel/exporters/learner_data.py#L113-L121)
- [where we use the course ID and user ID to get the learner data from the gradebook API](https://github.com/edx/edx-enterprise/blob/2931d8eab3f70539139a676468843a89cf4d09c1/integrated_channels/integrated_channel/exporters/learner_data.py#L558)
- [where we pass individual subsections of data back to the transmitter, and then subsequently to the client](https://github.com/edx/edx-enterprise/blob/master/integrated_channels/integrated_channel/exporters/learner_data.py#L127-L139)

Reply:
I'm guessing this was the comment you were telling me about lol. It's greatly informative and does bring up valid concerns that we should think about and maybe address with product. I agree that courses being sent to our integrations when we only ever associate grades with course runs is a bit of a problem.

To clarify the point I was initially bringing up: `I don't think an enterprise learner would be enrolled in courses not contained within the customer's catalog`
The scenario I was thinking of was not that the catalog was missing, but rather the scenario where a customer has multiple catalogs, but has only designated a subset to be sent to their integrated LMS. This is a really doubtful scenario and with the way I've configured the front end, the customer wouldn't be able to set this up themselves (I removed the catalog field, so it just sends all of their catalogs), but I just wanted to raise this possible edge case. We can also ignore it, since it is very unlikely to be hit - that is a perfectly good option.